### PR TITLE
ARROW-15300: [C++] Update Skyhook for async dataset interfaces

### PR DESCRIPTION
**Pull request by lidavidm**
After the synchronous dataset interfaces were removed in ARROW-13554, we need to update Skyhook to use the async interfaces. For now, we just use the same synchronous path as before, but on an I/O thread.

### CODE REVIEW

**Comment by lidavidm** [comment]
@github-actions crossbow submit -g test-skyhook-integration

**Comment by lidavidm** [comment]
@github-actions crossbow submit test-skyhook-integration

**Comment by lidavidm** [comment]
@github-actions crossbow submit test-skyhook-integration

**Comment by lidavidm** [comment]
@JayjeetAtGithub I'm trying to update this since we removed an API for 7.0.0, do you have any idea about the failure here? https://github.com/ursacomputing/crossbow/runs/4778154378?check_suite_focus=true

```
Running main() from /build/cpp/googletest_ep-prefix/src/googletest_ep/googletest/src/gtest_main.cc
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from TestSkyhookCLS
[ RUN      ] TestSkyhookCLS.SelectEntireDataset
[       OK ] TestSkyhookCLS.SelectEntireDataset (287 ms)
[ RUN      ] TestSkyhookCLS.SelectFewRows
[       OK ] TestSkyhookCLS.SelectFewRows (98 ms)
[ RUN      ] TestSkyhookCLS.SelectFewColumns
[       OK ] TestSkyhookCLS.SelectFewColumns (120 ms)
[ RUN      ] TestSkyhookCLS.SelectRowsAndColumnsOnPartitionKey
[       OK ] TestSkyhookCLS.SelectRowsAndColumnsOnPartitionKey (96 ms)
[ RUN      ] TestSkyhookCLS.SelectRowsAndColumnsOnlyOnPartitionKey
[       OK ] TestSkyhookCLS.SelectRowsAndColumnsOnlyOnPartitionKey (84 ms)
[----------] 5 tests from TestSkyhookCLS (687 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (688 ms total)
[  PASSED  ] 5 tests.
malloc_consolidate(): invalid chunk size
/arrow/ci/scripts/integration_skyhook.sh: line 135: 16922 Aborted                 (core dumped) debug/skyhook-cls-test
```

(I'm also having trouble with it locally, as it just hangs in TestSkyhookCLS.SelectEntireDataset, but that seems to be a problem with my own setup.)

**Comment by lidavidm** [comment]
Hmm, I can get it running locally now, interestingly that crash doesn't happen all the time. Under gdb, it appears this happens during thread shutdown…during which there's not much info left about what's going on, unfortunately.

AddressSanitizer does report a heap-use-after-free after test exit. I compared to the commit prior to 2e8b8365d602d8cb14a37ea4f349824549e28e8f and saw that it also has a heap-use-after-free, so it seems unrelated to what's going on here. (Or possibly, it's just that the changes here are just exposing it.)

**Comment by lidavidm** [comment]
@github-actions crossbow submit test-skyhook-integration

**Comment by lidavidm** [comment]
Ok, a little CMake tweak appears to fix it! Should be ready for review now.

**Comment by JayjeetAtGithub** [comment]
@lidavidm Thanks a lot for working on this. Sorry for the late reply. Let me take a look.

**Comment by JayjeetAtGithub** [comment]
Looks like the tests are passing. Please let me know if I can help with something. Thanks.

**Comment by lidavidm** [comment]
Ah, yes, I managed to fix the tests later on. Does the implementation look reasonable to you @JayjeetAtGithub? It's mostly the same code ported to the new interfaces.

**Review by westonpace — APPROVED** [review]
Thanks for getting to this!

File: cpp/src/skyhook/client/file_skyhook.cc [hunk 1]
```diff
@@ -95,10 +43,13 @@ class SkyhookFileFormat::Impl {
...
-  arrow::Result<arrow::dataset::ScanTaskIterator> ScanFile(
+  arrow::Result<arrow::RecordBatchGenerator> ScanBatchesAsync(
+      const std::shared_ptr<const SkyhookFileFormat>& format,
       const std::shared_ptr<arrow::dataset::ScanOptions>& options,
       const std::shared_ptr<arrow::dataset::FileFragment>& file) const {
-    /// Make sure client-side filtering and projection is turned off.
+    // TODO: investigate "true" async version
+
+    // Make sure client-side filtering and projection is turned off.
     file->apply_compute = false;
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/skyhook/client/file_skyhook.cc]: Ah, good point. I've gone and removed the flag. We can file a follow up for this case.

File: cpp/src/skyhook/client/file_skyhook.cc [hunk 2]
```diff
@@ -95,10 +43,13 @@ class SkyhookFileFormat::Impl {
     return arrow::Status::OK();
   }
 
-  arrow::Result<arrow::dataset::ScanTaskIterator> ScanFile(
+  arrow::Result<arrow::RecordBatchGenerator> ScanBatchesAsync(
+      const std::shared_ptr<const SkyhookFileFormat>& format,
       const std::shared_ptr<arrow::dataset::ScanOptions>& options,
       const std::shared_ptr<arrow::dataset::FileFragment>& file) const {
-    /// Make sure client-side filtering and projection is turned off.
+    // TODO: investigate "true" async version
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/skyhook/client/file_skyhook.cc]: You're right, I don't think there's much else we can do unless Ceph itself exposes an async interface for this.

**Comment by lidavidm** [comment]
@github-actions crossbow submit test-skyhook-integration

**Pull request closed without merging**